contracts/src/v0.8/dev/CanaryUpkeep.sol

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.13;
+
+import "../interfaces/KeeperCompatibleInterface.sol";
+import "../interfaces/KeeperRegistryInterface.sol";
+import "../ConfirmedOwner.sol";
+
+/**
+ * @notice A canary upkeep which requires a different keeper to service its upkeep at an interval. This makes sure that
+ * all keepers are in a healthy state.
+ */
+contract CanaryUpkeep is KeeperCompatibleInterface, ConfirmedOwner {
+  uint256 private s_keeperIndex;
+  uint256 private s_interval;
+  uint256 private s_timestamp;
+  KeeperRegistryInterface private immutable i_keeperRegistry;
+
+  /**
+   * @param keeperRegistry address of a keeper registry
+   */
+  constructor(KeeperRegistryInterface keeperRegistry, uint256 interval) ConfirmedOwner(msg.sender) {
+    i_keeperRegistry = keeperRegistry;
+    s_timestamp = block.timestamp;
+    s_interval = interval;
+    s_keeperIndex = 0;
+  }
+
+  /**
+   * @return the current keeper index
+   */
+  function getKeeperIndex() external view returns (uint256) {
+    return s_keeperIndex;
+  }
+
+  /**
+   * @return the current timestamp
+   */
+  function getTimestamp() external view returns (uint256) {
+    return s_timestamp;
+  }
+
+  /**
+   * @return the current interval
+   */
+  function getInterval() external view returns (uint256) {
+    return s_interval;
+  }
+
+  /**
+   * @notice updates the interval
+   * @param interval the new interval
+   */
+  function setInterval(uint256 interval) external onlyOwner {
+    s_interval = interval;
+  }
+
+  /**
+   * @notice returns true if keeper array is not empty and sufficient time has passed
+   */
+  function checkUpkeep(
+    bytes calldata /* checkData */
+  ) external view returns (bool, bytes memory) {
+    bool upkeepNeeded = block.timestamp >= s_interval + s_timestamp;
+    return (upkeepNeeded, bytes(""));
+  }
+
+  /**
+   * @notice checks keepers array limit, timestamp limit, and requires transaction origin must be the anticipated keeper.
+   * If all checks pass, update the keeper index and timestamp. Otherwise, revert this transaction.
+   */
+  function performUpkeep(
+    bytes calldata /* performData */
+  ) external {
+    (State memory _s, Config memory _c, address[] memory keepers) = i_keeperRegistry.getState();
+    if (keepers.length == 0) {
+      revert("no keeper nodes exists");
+    }
+    if (block.timestamp < s_interval + s_timestamp) {
+      revert("Not enough time has passed after the previous upkeep");
```

Comment:
can you use [custom errors instead of error strings](https://www.notion.so/chainlink/Solidity-Style-Guide-daa6418c56c04a45a2ff8c2655824c86#518dad1e9f9349c1b7099eaa58c71c10)

Reply:
Done